lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp

```diff
@@ -4981,10 +4990,12 @@ void SROA_Parameter_HLSL::flattenArgument(
               Twine("semantic '") + eltSem + "' on field overridden by function or enclosing type");
           }
 
-          // Inherit semantic from parent, but only preserve it for the first element.
+          // Inherit semantic from parent, but only preserve it for the first
+          // non-zero-sized element.
           // Subsequent elements are noted with a special value that gets resolved
           // once the argument is completely flattened.
-          EltAnnotation.SetSemanticString(i == 0 ? semantic : ContinuedPseudoSemantic);
+          EltAnnotation.SetSemanticString(
+            i == firstNonEmptyIx ? semantic : ContinuedPseudoSemantic);
```

Comment:
oh shoot! I merged this into the main branch, but I need to merge this change into hlsl2021. I forgot that this was in hlsl2021

Reply:
I've cherry-picked it into this PR already.

Reply:
yeah, but I don't want those conflicts when we do merge. I've created a merge here: #3850  sorry for leading you down this path only to draw things out 😣

Reply:
No problem, I'll rebase again.

Reply:
I've  updated the hlsl-2021 branch. You should be able to rebase now

Reply:
Done, but the Ubuntu build failed due to timeout.